dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.cpp

```diff
@@ -0,0 +1,415 @@
+        elapsed_since_schedule);
+}
+
+void LocalIndexerScheduler::moveBackReadyTasks(std::unique_lock<std::mutex> & lock)
+{
+    for (auto it = unready_tasks.begin(); it != unready_tasks.end();)
+    {
+        auto & task = *it;
+        if (isTaskReady(lock, task))
+        {
+            ready_tasks[task->user_task.keyspace_id][task->user_task.table_id].emplace_back(task);
+            it = unready_tasks.erase(it);
+        }
+        else
+        {
+            it++;
+        }
+    }
+}
+
+bool LocalIndexerScheduler::tryAddTaskToPool(std::unique_lock<std::mutex> & lock, const InternalTaskPtr & task)
+{
+    // Memory limit reached
+    if (pool_max_memory_limit > 0 && pool_current_memory + task->user_task.request_memory > pool_max_memory_limit)
+    {
+        return false;
+    }
+
+    auto real_job = [task, this]() {
+        SCOPE_EXIT({
+            std::unique_lock lock(mutex);
+            pool_current_memory -= task->user_task.request_memory;
+            running_tasks_count--;
+            taskOnFinish(lock, task);
+            on_finish_notifier.notify_all();
+
+            scheduler_need_wakeup = true;
+            scheduler_notifier.notify_all();
+        });
+
+        task->scheduled_at.start();
+
+        try
+        {
+            task->user_task.workload();
+        }
+        catch (...)
+        {
+            // TODO: We should ensure ADD INDEX task is retried somewhere else?
+            tryLogCurrentException(
+                logger,
+                fmt::format(
+                    "LocalIndexScheduler meet exception when running task: keyspace_id={} table_id={}",
+                    task->user_task.keyspace_id,
+                    task->user_task.table_id));
+        }
+    };
+
+    RUNTIME_CHECK(pool);
+    if (!pool->trySchedule(real_job))
+        // Concurrent task limit reached
+        return false;
+
+    ++running_tasks_count;
+    pool_current_memory += task->user_task.request_memory;
+    taskOnSchedule(lock, task);
+
+    return true;
+}
+
+LocalIndexerScheduler::ScheduleResult LocalIndexerScheduler::scheduleNextTask(std::unique_lock<std::mutex> & lock)
+{
+    if (ready_tasks.empty())
+        return ScheduleResult::FAIL_NO_TASK;
+
+    // Find the keyspace ID which is just > last_schedule_keyspace_id.
+    auto keyspace_it = ready_tasks.upper_bound(last_schedule_keyspace_id);
+    if (keyspace_it == ready_tasks.end())
+        keyspace_it = ready_tasks.begin();
+    const KeyspaceID keyspace_id = keyspace_it->first;
+
+    auto & tasks_by_table = keyspace_it->second;
+    RUNTIME_CHECK(!tasks_by_table.empty());
+
+    TableID last_schedule_table_id = -1;
+    if (last_schedule_table_id_by_ks.find(keyspace_id) != last_schedule_table_id_by_ks.end())
+        last_schedule_table_id = last_schedule_table_id_by_ks[keyspace_id];
+
+    auto table_it = tasks_by_table.upper_bound(last_schedule_table_id);
+    if (table_it == tasks_by_table.end())
+        table_it = tasks_by_table.begin();
```

Comment:
I agree that the scheduler should be fairly between different keyspaces.
But inside a keyspace, if the user wants to create vector index on multiple tables at the same time. Maybe letting one table finish its index first has a better user experience than all tables finish at almost the same time? @Lloyd-Pottiger @breezewish

Reply:
```cpp
auto table_it = tasks_by_table.find(last_schedule_table_id);
```